File: src/main/java/leetcode/hot100/LeetCode538.java
package leetcode.hot100;

import common.TreeNode;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/8 12:57
 */
public class LeetCode538 {

    int sum = 0;
    public TreeNode convertBST(TreeNode root) {
        if (root != null) {
            convertBST(root.right);
            sum += root.val;
            root.val = sum;
            convertBST(root.left);
        }
        return root;
    }
}

File: go/src/leetcode/hot100/LeetCode538.go
package hot100

func convertBST(root *TreeNode) *TreeNode {

	if root == nil {
		return nil
	}
	var dfs func(root *TreeNode)
	last := 0
	dfs = func(root *TreeNode) {
		if root == nil {
			return
		}
		dfs(root.Right)
		root.Val = last + root.Val
		last = root.Val
		dfs(root.Left)
		return
	}
	dfs(root)
	return root
}
